text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,262 @@
+- Feature Name: scoped
+- Start Date: 2015-04-16
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes an alternative to the `thread::scoped` API that does not rely
+on RAII, and is therefore memory safe even if destructors can be leaked.
+
+*This RFC was inspired by ideas from @nikomatsakis and @arielb1.*
+
+# Motivation
+
+The `thread::scoped` API allows parent threads to launch child threads that can
+have references to *stack data* owned by the parent. This style of API makes
+fork-join programming easy and efficient. It is also safe, *as long as the
+parent's stack frame outlives the children threads*.
+
+In the initial version of the API, safety was "guaranteed" through an RAII-style
+guard which, upon destruction, would join (wait for) any still-living children
+threads. This join guard is connected to the lifetime of any borrows by children
+threads, thereby (we thought) preventing the related parent stack frames from
+being popped until the joins were complete.
+
+Unfortunately,
+[it is possible to avoid the guard destructor in safe Rust](https://github.com/rust-lang/rust/issues/24292). Rust
+does not guarantee that the destructor for a value that escapes will be run.
+
+Note that this does *not* mean that RAII should be avoided in general. **RAII is
+still a safe, idiomatic, and ergonomic style of programming for Rust.** It just
+means that you cannot write `unsafe` code whose safety relies on RAII to run
+destructors.
+
+It may be possible to finesse this issue by locating every possible source of
+destructor leakage (such as `Rc` cycles) and tying it to some kind of "`Leak`"
+marker trait, which would be implemented by default for all Rust types (and
+basically means "safe to leak").  Ideas along these lines have been discussed
+extensively [elsewhere](https://github.com/rust-lang/rfcs/pull/1066), but they
+represent longer-term language changes that may or may not work out.
+
+This RFC, by contrast, proposes a general API that works in today's Rust, and
+can:
+
+* Guarantee execution of a piece of code prior to a scope being exited, even in
+  the presence of unwinding, `Rc` cycles, and so on.
+
+* Thereby be used to recover a safe `thread::scoped` API.
+
+While the API proposed here is slightly less ergonomic than the RAII-style
+`scoped` API, **it has the benefit of much more clearly marking the point at
+which children threads will be joined, which was left more implicit in the old
+API.**
+
+# Detailed design
+
+The proposal has two pieces: a general mechanism for "deferred computation",
+and a new `thread::scoped` API that takes advantage of it.
+
+## Deferred computation
+
+Here's a very simple API for deferring computation:
+
+```
+// in std:
+
+mod thread {
+    pub struct Scope<'a> { ... }
+
+    pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R;
```

Comment:
I wonder, could the lifetime parameter be omitted here as it's otherwise just anonymous?